### DISCUSSION TITLE
add scripting support

### DISCUSSION THREAD

**Discussion by khimaros** [maximecb/noisecraft#74] [Ideas]
BespokeSynth has a nice implementation of python scripting as a point of reference. i particularly love the live introspection ability (show the value of specific vars inline) and also highlighting which line is being executed.

it seems like javascript would be a nice companion for noisecraft :)

i'm sure this is a substantial effort, but thought it was worth throwing on the backlog.

**Comment by maximecb** [comment]
I've converted this to a discussion. Bespoke looks neat, but it's very complex and busy, it has a lot of modules, and each module has lots of functionality. That can be powerful, but it also makes it not so friendly for beginners IMO.

There's pros and cons in terms of scripting. Among the cons:
- Many of the people using NoiseCraft aren't programmers. If they open projects and find lots of programming code, they might find that off putting. Adding JS scripting means we're making the learning curve more steep and potentially making NoiseCraft more exclusive to a smaller niche of users.
- It adds complexity to an already complex system
- I don't think we'd want to allow full JS scripting because that would be insecure. We'd need to sandbox it somehow, but sandboxing things properly can be hard.

One thing that we could consider, though, is to have a node that allows people to write an expression, with multiple inputs and one output. The set of operators would be restricted, `+, -, *, /, sin, cos, tanh, etc`. That adds some of the functionality that you might find in scripting and allows making some things potentially more compact. Even then I'm a little hesitant. I don't really have support for configuring the number of inputs on a node right now and we'd need to program an expression parser.

**Comment by giiutfff** [comment]
Just an idea: expose only input, output and knobs then execute everything in a worker.

1. Non-programmers only see the interface, and they can reuse scripts already created.
2. Of course. But also more fun.
3. Workers have limited context(No DOM etc), and can only exchange data with main thread via `postMessage`. This solves some security issues. An expression parser is hardly turing complete, so maybe not enough for some cases.

Cons:
Exchange data between threads adds latency. Definitely not good for real time application.
